1_data_prep/13_tree_crops.py

```diff
@@ -0,0 +1,353 @@
+        None: If level is None (no mapping)
+    """
+    if level is None:
+        print("No species mapping specified - using original species codes")
+        return None
+    
+    if level not in ["l1", "l2", "l3", "l4"]:
+        raise ValueError(f"Invalid mapping level '{level}'. Must be one of: l1, l2, l3, l4, or None")
+    
+    # Load crosswalk
+    crosswalk_df = pd.read_csv(crosswalk_path)
+    
+    # Column names for the specified level
+    primary_col = f"primary_species_{level}"
+    species_col = f"species_code_{level}"
+    
+    # Create mapping dictionary: species_code -> target level species (only for primary_species_lX == True)
+    species_mapping = {}
+    for _, row in crosswalk_df.iterrows():
+        if row[primary_col]:  # Only include species where primary_species_lX is True
+            species_mapping[row['species_code']] = row[species_col]
+    
+    print(f"Loaded species crosswalk for level {level.upper()} with {len(species_mapping)} valid mappings")
+    print(f"Available {level.upper()} classes: {sorted(set(species_mapping.values()))}")
+    
+    return species_mapping
+
+def map_species(original_species, species_mapping, level):
+    """
+    Map original species to target level, handling unmapped species.
+    
+    Args:
+        original_species (str): Original species code
+        species_mapping (dict): Mapping dictionary (None if no mapping)
+        level (str): Target mapping level
+    
+    Returns:
+        tuple: (mapped_species, is_mapped)
+            - mapped_species: Target level species or original species if no mapping
+            - is_mapped: Boolean indicating if mapping was successful
+    """
+    if species_mapping is None:
+        return original_species, True  # No mapping requested, consider all as "mapped"
+    
+    if original_species in species_mapping:
+        return species_mapping[original_species], True
+    else:
+        return original_species, False
+
+# Load species mapping based on configuration
+species_mapping = load_species_mapping(species_crosswalk_path, MAPPING_LEVEL)
+
+# assemble list of dirs
+# ["0001_001435_001436", "0002_000451_000446", ...]
+dset_names = sorted(os.listdir(tree_label_mask_paths))
+dset_gt_mapping = {dset_name: gpd.read_file(Path(ground_data_path, dset_name+'.gpkg')) for dset_name in dset_names}
+
+# Process each dataset individually
+missing_img_ctr = 0
+skipped_datasets = []
+mapping_stats = {
+    'total_processed': 0,
+    'mapped_species': 0,
+    'unmapped_species': 0,
+    'saved_crops': 0,
+    'skipped_unmapped': 0
+}
+
+for dset_name in dset_names:
+    # metadata records for this specific dataset
+    records = []
+    
+    dset_idx, nadir_id, oblique_id = dset_name.split("_")
+
+    # Check if output folder for this dataset already exists in 'labelled'
+    labelled_output_dir = Path(output_path, "labelled", dset_name)
+    if labelled_output_dir.exists():
```

Comment:
Can we check that the metadata file has been created, rather than the top level folder? That would ensure all the crops have been generated, which could be helpful if it dies in the middle of a given dataset.